src/components/app/custom-els/FileDrop/index.ts

```diff
@@ -0,0 +1,124 @@
+import * as styles from './styles.css';
+import { bind } from '../../../../lib/util';
+
+function convertMimeTypeToRegex(mimeType: string): string {
+  return mimeType.replace(/\*/, '[^/]+').replace(/\//g, '\\/');
+}
+
+function firstMatchingFile(list: DataTransferItemList, accept: string): DataTransferItem | undefined {
+  const mimeRegex = convertMimeTypeToRegex(accept);
+  const acceptsRegex = new RegExp(mimeRegex);
+
+  return Array.from(list).find(item => item.kind === 'file' && acceptsRegex.test(item.type));
+}
+
+interface FileDropEventInit extends EventInit {
+  file: File;
+}
+
+export class FileDropEvent extends Event {
+  private _data: FileDropEventInit;
+  constructor(typeArg: string, eventInitDict: FileDropEventInit) {
+    super(typeArg, eventInitDict);
+    this._data = eventInitDict;
+  }
+
+  get data (): FileDropEventInit {
+    return this._data;
+  }
+}
```

Comment:
You also need:

```ts
interface HTMLElementEventMap {
  'filedrop': FileDropEvent;
}
```

This means TypeScript will know that listeners to "filedrop" will receive a `FileDropEvent`.

Reply:
hmm - I'm not sure how to reference this at all.... in missing types, or in the main file of the class.

Reply:
I think I worked it out.